src/transformers/modeling_t5.py

```diff
@@ -948,16 +1104,34 @@ def forward(
             # get decoder inputs from shifting lm labels to the right
             decoder_input_ids = self._shift_right(lm_labels)
 
+        # If decoding with past key value states, only the last tokens
+        # should be given as an input
+        if decoder_past_key_value_states is not None and self.decoder.output_past is True:
+            assert (
+                lm_labels is None
+            ), "Decoder should not use cached key value states when training. Also consider setting model.set_output_past(False) for less memory consumption"
+            if decoder_input_ids is not None:
+                decoder_input_ids = decoder_input_ids[:, -1:]
+            if decoder_inputs_embeds is not None:
+                decoder_inputs_embeds = decoder_inputs_embeds[:, -1:]
+
         # Decode
         decoder_outputs = self.decoder(
             input_ids=decoder_input_ids,
             attention_mask=decoder_attention_mask,
             inputs_embeds=decoder_inputs_embeds,
+            past_key_value_states=decoder_past_key_value_states,
             encoder_hidden_states=hidden_states,
             encoder_attention_mask=attention_mask,
             head_mask=head_mask,
         )
 
+        # insert decoder past at right place
+        # to speed up decoding
+        if self.decoder.output_past:
+            past = ((encoder_outputs, decoder_outputs[1]),)
+            decoder_outputs = decoder_outputs[:1] + past + decoder_outputs[2:]
```

Comment:
That's also important. As in done in Bart the past (`output[1]` in the generate fn) contains both the `encoder_outputs` and the cached key value states. This allows us to only work with the `past` variable in all `prepare_inputs_for_generation` functions, but it requires to have two pointers to the encoder outputs in all outputs for T5.